cns/deviceplugin/server.go

```diff
@@ -0,0 +1,153 @@
+			conn, err := (&net.Dialer{}).DialContext(ctx, "unix", addr)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to dial context")
+			}
+			return conn, nil
+		}),
+	)
+	if err != nil {
+		return errors.Wrap(err, "error dialing local grpc server")
+	}
+	if err := c.Close(); err != nil {
+		return errors.Wrap(err, "error closing connection to local grpc server")
+	}
+	return nil
+}
+
+func (s *Server) Allocate(_ context.Context, req *v1beta1.AllocateRequest) (*v1beta1.AllocateResponse, error) {
+	s.logger.Info("allocate request", zap.Any("req", *req))
+	resps := make([]*v1beta1.ContainerAllocateResponse, len(req.ContainerRequests))
+	for i, containerReq := range req.ContainerRequests {
+		resp := &v1beta1.ContainerAllocateResponse{
+			Envs: make(map[string]string),
+		}
+		for j := range containerReq.DevicesIDs {
+			resp.Envs[fmt.Sprintf("NIC_%d", j)] = containerReq.DevicesIDs[j]
+		}
+		resps[i] = resp
+	}
+	r := &v1beta1.AllocateResponse{
+		ContainerResponses: resps,
+	}
+	return r, nil
+}
+
+func (s *Server) ListAndWatch(_ *v1beta1.Empty, stream v1beta1.DevicePlugin_ListAndWatchServer) error {
+	// send the initial count right away
+	advertisedCount := s.deviceCounter.getDeviceCount()
+	devices := make([]*v1beta1.Device, advertisedCount)
+	for i := range devices {
+		devices[i] = &v1beta1.Device{
+			ID:     fmt.Sprintf("NIC-%d", i),
+			Health: v1beta1.Healthy,
+		}
+	}
+	if err := stream.Send(&v1beta1.ListAndWatchResponse{
+		Devices: devices,
+	}); err != nil {
+		return errors.Wrap(err, "error sending listAndWatch response")
+	}
+
+	// every interval, check if the current count has changed from what we've previously sent, and if so, send the new count
+	for {
+		select {
+		case <-s.listAndWatchDoneChan:
+			return nil
+		case <-time.After(s.deviceCheckInterval):
+			currentCount := s.deviceCounter.getDeviceCount()
+			if currentCount == advertisedCount {
+				continue
+			}
+			advertisedCount = currentCount
+			devices := make([]*v1beta1.Device, advertisedCount)
+			for i := range devices {
+				devices[i] = &v1beta1.Device{
+					ID:     fmt.Sprintf("vnet-nic-%d", i),
+					Health: v1beta1.Healthy,
+				}
+			}
+			if err := stream.Send(&v1beta1.ListAndWatchResponse{
+				Devices: devices,
+			}); err != nil {
+				return errors.Wrap(err, "error sending listAndWatch response")
+			}
+		}
+	}
+}
+
+func (s *Server) GetDevicePluginOptions(context.Context, *v1beta1.Empty) (*v1beta1.DevicePluginOptions, error) {
+	return &v1beta1.DevicePluginOptions{}, nil
+}
+
+func (s *Server) GetPreferredAllocation(context.Context, *v1beta1.PreferredAllocationRequest) (*v1beta1.PreferredAllocationResponse, error) {
+	return &v1beta1.PreferredAllocationResponse{}, nil
+}
+
+func (s *Server) PreStartContainer(context.Context, *v1beta1.PreStartContainerRequest) (*v1beta1.PreStartContainerResponse, error) {
+	return &v1beta1.PreStartContainerResponse{}, nil
+}
```

Comment:
indeed, it's the v1beta1.DevicePluginServer I believe (have to double check the type)

Reply:
Thats correct.
```
type DevicePluginServer interface {
	// GetDevicePluginOptions returns options to be communicated with Device
	// Manager
	GetDevicePluginOptions(context.Context, *Empty) (*DevicePluginOptions, error)
	// ListAndWatch returns a stream of List of Devices
	// Whenever a Device state change or a Device disappears, ListAndWatch
	// returns the new list
	ListAndWatch(*Empty, DevicePlugin_ListAndWatchServer) error
	// GetPreferredAllocation returns a preferred set of devices to allocate
	// from a list of available ones. The resulting preferred allocation is not
	// guaranteed to be the allocation ultimately performed by the
	// devicemanager. It is only designed to help the devicemanager make a more
	// informed allocation decision when possible.
	GetPreferredAllocation(context.Context, *PreferredAllocationRequest) (*PreferredAllocationResponse, error)
	// Allocate is called during container creation so that the Device
	// Plugin can run device specific operations and instruct Kubelet
	// of the steps to make the Device available in the container
	Allocate(context.Context, *AllocateRequest) (*AllocateResponse, error)
	// PreStartContainer is called, if indicated by Device Plugin during registeration phase,
	// before each container start. Device plugin can run device specific operations
	// such as resetting the device before making devices available to the container
	PreStartContainer(context.Context, *PreStartContainerRequest) (*PreStartContainerResponse, error)
}
```